[FLINK-33728] Do not rewatch when KubernetesResourceManagerDriver wat…

zhougit86 opened the pull request:
…ch fail



## What is the purpose of the change

prevent jm restart in case kubernetes fails to handle watch request


## Brief change log

use ExponentialBackoffRetryStrategy to retry watch


## Verifying this change


This change is a trivial rework / code cleanup without any test coverage.


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: ( no )
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes )
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs)

zhougit86 commented:
@xintongsong Would you please take a look and let me know your thinking

xintongsong commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java:
```diff
@@ -440,6 +441,28 @@ public class KubernetesConfigOptions {
+    public static final ConfigOption<Duration> KUBERNETES_TRANSACTIONAL_OPERATION_INITIAL_RETRY_DEALY =
+            key("kubernetes.transactional-operation.initial-retry-delay")
+                    .durationType()
+                    .defaultValue(Duration.ofMillis(5000))
```
I think the 5s default initial delay might be too large. I'd suggest maybe 50ms.

xintongsong replied:
That also means it would take 12 retries to reach the default 1m max delay. Maybe we can also increase the default value for max-retires to maybe 15 or 20 times?

xintongsong commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java:
```diff
@@ -102,6 +109,14 @@ public Fabric8FlinkKubeClient(
         this.maxRetryAttempts =
                 flinkConfig.getInteger(
                         KubernetesConfigOptions.KUBERNETES_TRANSACTIONAL_OPERATION_MAX_RETRIES);
+        this.initialRetryInterval =
+                flinkConfig.getOptional(
+                                KubernetesConfigOptions.KUBERNETES_TRANSACTIONAL_OPERATION_INITIAL_RETRY_DEALY)
+                        .orElse(KubernetesConfigOptions.KUBERNETES_TRANSACTIONAL_OPERATION_INITIAL_RETRY_DEALY.defaultValue());
+        this.maxRetryInterval =
+                flinkConfig.getOptional(
+                                KubernetesConfigOptions.KUBERNETES_TRANSACTIONAL_OPERATION_MAX_RETRY_DEALY)
+                        .orElse(KubernetesConfigOptions.KUBERNETES_TRANSACTIONAL_OPERATION_MAX_RETRY_DEALY.defaultValue());
```
There's no need to use `getOptional().orElse()`. A simple `get()` should automatically apply the default value if not configured.

```suggestion
        this.initialRetryInterval =
                flinkConfig.get(
                                KubernetesConfigOptions.KUBERNETES_TRANSACTIONAL_OPERATION_INITIAL_RETRY_DEALY);
        this.maxRetryInterval =
                flinkConfig.get(
                                KubernetesConfigOptions.KUBERNETES_TRANSACTIONAL_OPERATION_MAX_RETRY_DEALY);
```

xintongsong reviewed (changes requested):
Thanks for working on this, @zhougit86.

I left some inline comments, PTAL.

In addition to the inline comment:
1. I think we should add some new test cases to verify and guard the newly introduced feature behavior.
2. There's some code-style violations that leads to CI failures. Please take a look.

zhougit86 commented:
@xintongsong Hi , I got a chance to modify my commit, would you please bother to review for another time?

xintongsong commented on docs/layouts/shortcodes/generated/kubernetes_config_configuration.html:
```diff
@@ -290,5 +290,17 @@
+            <td style="word-wrap: break-word;">5</td>
+            <td>Duration</td>
+            <td>Defines the max duration of Kubernetes transactional operation retries after fail.</td>
+        </tr>
```
We need to re-generate this document. The CI failure is related to this.

xintongsong commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesClientTestBase.java:
```diff
@@ -183,6 +183,29 @@ protected void mockPodEventWithLabels(
     }
 
+    protected void mockPodEventWithLabelsFail(
+            String namespace, String podName, String resourceVersion, Map<String, String> labels) {
```
`podName` is never used.

xintongsong commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesClientTestBase.java:
```diff
@@ -183,6 +183,29 @@ protected void mockPodEventWithLabels(
 
+    protected void mockPodEventWithLabelsFail(
+            String namespace, String podName, String resourceVersion, Map<String, String> labels) {
+        final Pod pod =
+                new PodBuilder()
+                        .withNewMetadata()
+                        .withNamespace(namespace)
+                        .withName(podName)
+                        .withLabels(labels)
+                        .withResourceVersion("5668")
+                        .endMetadata()
+                        .build();
```
It seems there's no need to create this `pod`.

xintongsong commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesClientTestBase.java:
```diff
@@ -183,6 +183,29 @@ protected void mockPodEventWithLabels(
                 .once();
     }
 
+    protected void mockPodEventWithLabelsFail(
```
This name does not reflect what this method really does. I'd suggest `mockWatchPodFailure`

xintongsong commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClientTest.java:
```diff
@@ -454,6 +451,22 @@ void testWatchPodsAndDoCallback() throws Exception {
+        TestingWatchCallbackHandler<KubernetesPod> watchCallbackHandler =
+                TestingWatchCallbackHandler.<KubernetesPod>builder().build();
+        Long watchStartTime = System.currentTimeMillis();
+        assertThatThrownBy(
+                () ->
+                        this.flinkKubeClient.watchPodsAndDoCallback(
+                                TESTING_LABELS, watchCallbackHandler));
```
It is unclear to me whether `watchPodsAndDoCallback` suppose to succeed or fail? There's no verification against the error being thrown.

xintongsong commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClientTest.java:
```diff
@@ -454,6 +451,22 @@ void testWatchPodsAndDoCallback() throws Exception {
+                NAMESPACE, TASKMANAGER_POD_NAME, KUBERNETES_ZERO_RESOURCE_VERSION, TESTING_LABELS);
+        TestingWatchCallbackHandler<KubernetesPod> watchCallbackHandler =
+                TestingWatchCallbackHandler.<KubernetesPod>builder().build();
+        Long watchStartTime = System.currentTimeMillis();
```
I'd suggest to use `System.nanoTime()`.

`System.currentTimeMillis()` can be updated at a larger granularity and is not suitable ms-level time counting.

xintongsong commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClientTest.java:
```diff
@@ -454,6 +451,22 @@ void testWatchPodsAndDoCallback() throws Exception {
+                                TESTING_LABELS, watchCallbackHandler));
+        Long watchEndTime = System.currentTimeMillis();
+        Long watchDuration = watchEndTime - watchStartTime;
+        assertThat(watchDuration > 100);
```
Relying on the duration to decide whether the retry has happened can be unstable. I'd suggest to mock the server in a way that replies failure messages for certain times following by a success message.

xintongsong commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClientTest.java:
```diff
@@ -454,6 +451,22 @@ void testWatchPodsAndDoCallback() throws Exception {
     }
 
+    @Test
+    void testWatchPodsAndDoCallbackFail() throws Exception {
```
`Exception` is never thrown.

xintongsong commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java:
```diff
@@ -247,11 +259,12 @@ public KubernetesWatch watchPodsAndDoCallback(
                                         .isPresent(),
-                        kubeClientExecutorService)
+                        new ScheduledExecutorServiceAdapter(kubeClientExecutorService))
                 .get();
```
This means `watchPodsAndDoCallback` is blocking and can wait for minutes to return. It should not longer be called on the RPC main thread.

xintongsong commented:
@zhougit86 ,
If you need any help in understanding the comments, feel free to reach out to my email (tonysong820@gmail.com). We may exchange WeChat ID and have an offline phone call if you'd like to.

xintongsong commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/TestingFlinkKubeClient.java:
```diff
@@ -149,9 +149,10 @@ public List<KubernetesPod> getPodsWithLabels(Map<String, String> labels) {
             Map<String, String> labels, WatchCallbackHandler<KubernetesPod> podCallbackHandler) {
-        return watchPodsAndDoCallbackFunction.apply(labels, podCallbackHandler);
+        return CompletableFuture.supplyAsync(
+                () -> watchPodsAndDoCallbackFunction.apply(labels, podCallbackHandler));
```
I think we should change the return value type of `watchPodsAndDoCallbackFunction` to align with the method signature.

zhougit86 commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java:
```diff
@@ -114,7 +116,8 @@ public KubernetesResourceManagerDriver(
     protected void initializeInternal() throws Exception {
-        podsWatchOpt = watchTaskManagerPods();
+        podsWatchOptFuture = watchTaskManagerPods();
+        podsWatchOptFuture.get();
```
This is because the watchTaskManagerPods before is a sync method, I just want to keep it the same behavior as before.

zhougit86 commented:
@flinkbot run azure

xintongsong commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java:
```diff
@@ -90,7 +92,7 @@ public class KubernetesResourceManagerDriver
     private long currentMaxPodId = 0;
 
-    private Optional<KubernetesWatch> podsWatchOpt;
+    private CompletableFuture<KubernetesWatch> podsWatchOptFuture;
```
```suggestion
    private CompletableFuture<KubernetesWatch> podsWatchOptFuture =
            FutureUtils.completedExceptionally(
                    new ResourceManagerException(
                            "KubernetesResourceManagerDriver is not initialized."));
```

xintongsong replied:
This helps prevent NPE if `podsWatchOptFuture` is accidentally being used before being initialized.

xintongsong commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java:
```diff
@@ -450,17 +462,30 @@ public void handleError(Throwable throwable) {
+                                                        }
+                                                    }
+                                                },
+                                                getMainThreadExecutor()));
```
The outer `getMainThreadExecutor().execute()` is unnecessary.

xintongsong reviewed (approved):
@zhougit86,

Thanks for addressing my comments. The PR LGTM.

The CI failures seem unrelated to this PR. I have just rebased the PR onto the latest master branch. Let's see if that resolves the CI failures.

I also have another two minor comments, which I addressed myself in a fixup commit. Please check them out.

If you have no objections on my changes, I'll merge this PR once it passes the CI.

zhougit86 commented:
> @zhougit86,
> 
> Thanks for addressing my comments. The PR LGTM.
> 
> The CI failures seem unrelated to this PR. I have just rebased the PR onto the latest master branch. Let's see if that resolves the CI failures.
> 
> I also have another two minor comments, which I addressed myself in a fixup commit. Please check them out.
> 
> If you have no objections on my changes, I'll merge this PR once it passes the CI.

Thanks for your advice, really appreciate!
